### DISCUSSION TITLE
Installation OK, but Fronius Powermon still not responding

### DISCUSSION THREAD

**Discussion by Thueringer-2** [bulletmark/corsproxy#3] [Q&A]
Dear Mr. Blakeney, please excuse my disturbing you. I stumbled across this project after I spent several days in investigating the question, why my Fronius inverter (Symo GEN24 6.0) refused to answer the request for real time data via `http://fronius-123456/solar_api/v1/GetPowerFlowRealtimeData.fcgi?Scope=Device&DeviceId=0` by Javascript. So I ended up with the CORS diagnosis as well, found your Github project and tried to install, but I must admit that I did not manage to get it to run. I am not too much into the depths of unixoid operating systems and came to a point where I kindly ask for advice. Here is what happened before.

- Followed installation procedure as per https://github.com/bulletmark/corsproxy
- Installed as user "pi" (same as default user on system) on a RPi4. 
- During setup, I configured
`~/.config/corsproxy `
with one line like this:
`8001:192.168.178.136:80`
- Furthermore, 
`/etc/systemd/system/corsproxy.service`
is configured like this:
```
[Unit]
Description=CORS Proxy Service
[Service]
User=pi
# Set the following to the source program directory (where you git cloned)
WorkingDirectory=/home/pi/corsproxy
ExecStart=/home/pi/corsproxy/%p
[Install]
WantedBy=default.target
```
- After reboot, 
`journalctl -u corsproxy `
yields:
```
Feb 06 19:16:56 MYMACHINE systemd[1]: Started CORS Proxy Service.
Feb 06 19:17:01 MYMACHINE corsproxy[387]: Starting proxy for 8001 -> 192.168.178.136:80
```
- So, all seems well.

**BUT**
- When I call 192.168.178.136 from a Chromium browser on the same machine, I get the Fronius welcome page. OK.
- When I call 192.168.178.136:8001, I would expect the same page to be loaded in the same way, but the request times out.
- When I try your Fronius power monitor tool (http://fronius-powermon.duckdns.org/) on the same machine and set the host under 'Settings' to `http://192.168.178.136:8001`, I also don't get any output.

Would you please be so kind as to give a hint where I should look for the source of error? Any help is greatly appreciated.

Thank you!

**Comment by Thueringer-2** [comment]
Fault finding so far:

`systemctl status corsproxy`
is responded with
```
corsproxy.service - CORS Proxy Service
   Loaded: loaded (/etc/systemd/system/corsproxy.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2023-02-06 19:16:56 CET; 21h ago
 Main PID: 387 (corsproxy)
    Tasks: 1 (limit: 4915)
   CGroup: /system.slice/corsproxy.service
           └─387 venv/bin/python -u /home/pi/corsproxy/corsproxy
Feb 06 19:16:56 MYSYSTEM systemd[1]: Started CORS Proxy Service.
Feb 06 19:17:01 MYSYSTEM corsproxy[387]: Starting proxy for 8001 -> 192.168.178.136:80
```

But nevertheless
`sudo traceroute -T -p 8001 192.168.178.136`
does not find a path (to port 80 it does).

**Comment by bulletmark** [comment]
Let's forget about corsproxy and your RPI4 at the moment and establish that a Fronius GEN24 provides the API needed. Please point your browser to http://192.168.187.80/solar_api/v1/GetPowerFlowRealtimeData.fcgi and confirm that it returns the JSON response needed. I assume here the IP address of your inverter is 192.168.187.80, please change that if I assumed wrongly.

**Reply by Thueringer-2** [reply]
Yes, this is working nicely, from any computer in the local network, including the RPi where corsproxy is running. 
Also, when I explicitly specify port 80 (http://192.168.178.136:80/solar_api/v1/GetPowerFlowRealtimeData.fcgi). 
It does not work when I specify port 8001 on the RPi (http://192.168.178.136:8001/solar_api....)
When I look for running webservers on the RPi by
`netstat -tlnp`
it says (Among others)
`tcp        0      0 0.0.0.0:8001            0.0.0.0:*               LISTEN      387/venv/bin/python `

**Comment by bulletmark** [comment]
Ted, please answer the following:

1. What is the IP address of your inverter?
2. What is the IP address of your RPi?
3. Please paste the output here of http://<IP_addr_of_inverter>/solar_api/v1/GetPowerFlowRealtimeData.fcgi

**Reply by Thueringer-2** [reply]
```
1. 192.168.178.136
2. 192.168.178.32
3. {
   "Body" : {
      "Data" : {
         "Inverters" : {
            "1" : {
               "Battery_Mode" : "nearly depleted",
               "DT" : 1,
               "E_Day" : null,
               "E_Total" : 5520239.4694444444,
               "E_Year" : null,
               "P" : -42.650310516357422,
               "SOC" : 20.0
            }
         },
         "SecondaryMeters" : {},
         "Site" : {
            "BackupMode" : false,
            "BatteryStandby" : false,
            "E_Day" : null,
            "E_Total" : 5520239.4694444444,
            "E_Year" : null,
            "Meter_Location" : "grid",
            "Mode" : "bidirectional",
            "P_Akku" : -0.14469008147716522,
            "P_Grid" : 371.69999999999999,
            "P_Load" : -329.04968948364257,
            "P_PV" : 4.9869224312715232e-05,
            "rel_Autonomy" : 0.0,
            "rel_SelfConsumption" : 100.0
         },
         "Smartloads" : {
            "Ohmpilots" : {}
         },
         "Version" : "12"
      }
   },
   "Head" : {
      "RequestArguments" : {},
      "Status" : {
         "Code" : 0,
         "Reason" : "",
         "UserMessage" : ""
      },
      "Timestamp" : "2023-02-09T05:14:53+00:00"
   }
}
```

**Comment by bulletmark** [comment]
Your configuration seems correct. Run `corsproxy` on the RPi with configuration `8001:192.168.176.136`. Then on the settings page on http://fronius-powermon.duckdns.org/ configure the host as `192.168.1.32:8001`. I.e. you point that page to your RPi proxy which fetches the inverter data on behalf of your browser and returns it.

**Reply by Thueringer-2** [reply]
Ah. My bad. It works like a charm now. 
I was indeed wrong about the function of a proxy. I understood it as sort of a port listener on the RPi. Now it is all clear and working.
The key point is that one does not call the inverter on port 8000 or so, but the system on which the proxy is running.
Thank you again for pointing me to the right direction, and as I know see your Fronius monitor working as well ... I am really excited about it! This is a very handy tool to monitor household appliances which are normally difficult to track. The Fronius app has a way too rough time scheme.
